metastore/metastore-api/src/main/java/org/apache/drill/metastore/util/SchemaPathUtils.java

```diff
@@ -63,6 +64,30 @@ public static ColumnMetadata getColumnMetadata(SchemaPath schemaPath, TupleMetad
     return colMetadata;
   }
 
+  /**
+   * Checks if field indetified by the schema path is child in either {@code DICT} or {@code REPEATED MAP}.
+   * For such fields, nested in {@code DICT} or {@code REPEATED MAP},
+   * filters can't be removed based on Parquet statistics.
+   * @param schemaPath schema path used in filter
+   * @param schema schema containing all the fields in the file
+   * @return {@literal true} if field is nested inside {@code DICT} (is {@code `key`} or {@code `value`})
+   *         or inside {@code REPEATED MAP} field, {@literal false} otherwise.
+   */
+  public static boolean isFieldNestedInDictOrRepeatedMap(SchemaPath schemaPath, TupleMetadata schema) {
```

Comment:
This is a kind of cluttered version of what was just suggested. Rather than a bunch of if-statements that answers a specific question, generalize to look up a member. You'd get the same result as this method if `member(String name)` returned `null` if no such member existed. (And, you'd do it generically without having to later extend this to `isFieldNestedInDictOrRepeatedMapOrSomeNewType()`.

Also, why only `Repeated` Map? (Non-repeated) Maps also have fields and allow `a.b` notation.

Reply:
Another general observation is that we've accidentally created multiple versions of the same code. The `RequestedTupleImpl` class does something very similar: it converts a `SchemaPath` into a consolidated projection set using logic much like we have here. For example, `SELECT a, a.b` is consolidated into a single column, `a`, that must be a MAP that must contain at least a `b` member. That code was very tricky to get right, as, I'm sure, this is. That code has to be modified to handle `SELECT a, a.key`. Is `a` now a `MAP` or a `DICT`? We don't know, all we know is that `a` is consistent with either interpretation.

Do we really need multiple copies?

Some things to consolidate:

* `ColumnMetadata` - which should be our go-to solution as it is the most general.
* `MaterializedField` - which has all kinds of holes and problems for complex types.
* `SerializedField` - Like `MaterializedField` but with buffer lengths?
* `RequestedColumn` - Part of that code mentioned above that parses a project list into a consolidated set of columns.
* `SchemaPath` - A description of a project list.
* The logic here
* ... - Probably others.

We probably want three tiers of representation:

* Project list - `SchemaPath`
* "Semanticized" project list - `RequestedTuple`
* Column metadata - `ColumnMetadata`

Then, we create one set of code that handles transforms and validations. (Convert project list to semantisized list, then validate that against a schema.) As it turns out, that is what the EVF does for scan projection planning; maybe we can pull that out and generalize it for use elsewhere?

Reply:
This method is used to check whether a schema path in filter, e.g. `... WHERE mapcol['a'] IS NULL`, references a `DICT`'s `value` (accessed by some key). If a `value` is `OPTIONAL INT` (which is default type for absent column also), first statistics is going to be retrieved for the 'field' identified by schema path. In case when `value` is retrieved by key, previous example results to schema path being `` `mapcol`.`a` `` which is not present in statistics (but there is statistics for the value itself, which has schema path `` `mapcol`.`map`.`value` ``) and then is treated as an absent column resulting in every row matching the filter.

While doing the changes for resolving correct metadata for the `DICT`, I've seen that `DataMode` for `key` and `value` fields are always `REPEATED`. This was because data mode was determined using Parquet's max `repetition` and `definition` level values (see changes in `ParquetTableMetadataUtils.java`), computed for the whole schema up to the leaf field. The algorith was, if `repetition >= 1` then the field is `REPEATED`. This means, if there is at least one `REPEATED` member in schema, each of its children is going to be `REPEATED`. In case of `DICT`, it has a nested `repeated group`, mentioned above, thus resulting in `key` and `value` being `REPEATED`. To retain original data mode, a Parquet's `Type.Repetition` was added to column metadata v4 and is used instead. This filtering was working before for the `DICT` because the `value` was `REPEATED` and statistics was not used for `REPEATED` fields. Now, when the data mode is retained, there is a need to handle such a case. `REPEATED MAP` is included, because its fields were determined to be `REPEATED` also, thus to preserve previous behaviuor. But currently, this method is likely to be used if the `REPEATED MAP` contains an `OPTIONAL INT`. For the case of repeated map, it's statistics is not found, because it uses simple name, e.g. `` `struct_array`.`a` `` (note that indexes are ommited, as they are not retained in Parquet schema), but it has another _actual_ structure: `` `struct_array`.`bag`.`array_element`.`a` ``.